### ISSUE TITLE
Updater not accessible for screen readers

### ISSUE THREAD

**Issue by sukiletxe** [git-for-windows/git#2099]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.20.1.windows.1
cpu: x86_64
built from commit: 7c9fbc07db0e2939b36095df45864b8cda19b64f
sizeof-long: 4
sizeof-size_t: 8

```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Versión 6.1.7601]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Notepad++
Custom Editor Path: 
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I use [NVDA](https://nvaccess.org) as my screen reader.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
N/A
```
 - What did you expect to occur after running these commands?

N/A

 - What actually happened instead?

N/A

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

** insert URL here **

Seeing as the build-extra repository doesn't have any issue facility, I post this here. Issues I'm seeing (figuratively, thugh clarification probably wasn't needed):
* When the updater launches, everything there are is non-standard controls. Everything my screen reader says is "unknown", which means it detects some objects, but it can't identify its control type nor text. I'm not sure about how this window was constructed, or even if any toolkit has been used. QT4, not sure about 5, and GTK are famous for providing these on Windows, whereas WX works fine most of the time for a dialog box of this nature. 
* Whenever I hit the tab key, I get no feedback (are these objects focusable with a keyboard?). I must access these by using a method called object navigation, which basically lays out the objects in a hierarchical fashion (that is, the application is the parent of four objects, none of which have children). I don't know how to explain this better, hope it makes sense.
* If I try directly navigating through the objects of the window and OCRing each of them, only the last one contains some text. This object is the one with the dialog box' text.

I have zero experience with GUIs, but perhaps if you can tell me how to trigger that dialog box again, I can provide you with some more information with the help of NVDA's log (but I have no idea if that will be useful).

**Comment by kgybels** [comment]
I'm assuming with updater you mean the dialog that pops up to notify that there is a new version to install.

When you enable the daily update check in the installer, it schedules a task to run daily to execute the following [command](https://github.com/git-for-windows/build-extra/blob/cad7897f73f1ac1810161772ce6c62986c9e1744/installer/install.iss#L2256-L2257): `"C:\Program Files\Git\git-bash.exe" --hide --no-needs-console --command=cmd\git.exe update-git-for-windows --quiet --gui`

Since you're on Windows 7, update-git-for-windows will [show the dialog](https://github.com/git-for-windows/build-extra/blob/cad7897f73f1ac1810161772ce6c62986c9e1744/git-extra/git-update-git-for-windows#L240-L241) with: `git gui--askyesno --title "Git Update Available" "Download and install Git for Windows 2.21.0?"`

[gui--askyesno](https://github.com/git-for-windows/git/blob/v2.21.0.windows.1/git-gui/git-gui--askyesno) uses Tcl/Tk to show the dialog, which doesn't seem to have accessibility support.

A possible solution would be to replace the current gui--askyesno script with a simple windows application that uses the [MessageBox function](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-messagebox), which I'm assuming has correct accessibility support.

**Comment by dscho** [comment]
> [gui--askyesno](https://github.com/git-for-windows/git/blob/v2.21.0.windows.1/git-gui/git-gui--askyesno) uses Tcl/Tk to show the dialog, which doesn't seem to have accessibility support.

That is too bad! We used it for pretty much everything :-(

> A possible solution would be to replace the current gui--askyesno script with a simple windows application that uses the [MessageBox function](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-messagebox), which I'm assuming has correct accessibility support.

Indeed, I think this is the way to go.

@sukiletxe could I ask you to verify that `git gui--askyesno "Yes or No?"` looks exactly as "unknown" to your screen reader as the updater?

**Comment by sukiletxe** [comment]
@kgybels :
> A possible solution would be to replace the current gui--askyesno script with a simple windows application that uses the [MessageBox function](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-messagebox), which I'm assuming has correct accessibility support.

Yes, that works. I assumed it did (after all it's from Microsoft and I use them almost daily), but after testing them with a simple Python program which uses the MFC class I confirmed it.

@dscho :
> @sukiletxe could I ask you to verify that `git gui--askyesno "Yes or No?"` looks exactly as "unknown" to your screen reader as the updater?

Exactly same layout as before, yes. Four unknown objects as children of the window.

**Comment by dscho** [comment]
@sukiletxe thank you for clarifying.

How are your C skills? If they are decent to good, it would be awesome if you could work on that `MessageBoxW()`- based `askyesno` helper.

I could imagine that a good place for it to live would be in the `git-extra` package, next to [`blocked-file-util.c`](https://github.com/git-for-windows/build-extra/blob/master/git-extra/blocked-file-util.c)

It would have the exact same command line like [`git gui--askyesno`](https://github.com/git-for-windows/git/blob/v2.21.0.windows.1/git-gui/git-gui--askyesno): taking a single argument that is the question to ask, with an optional `--title <title>` argument before it, and exiting with code 0 in case of Yes and 1 in case of No.

It would then have to be [copied like `blocked-file-util.c`](https://github.com/git-for-windows/build-extra/blob/cad7897f73f1ac1810161772ce6c62986c9e1744/git-extra/PKGBUILD#L48), and [built like `blocked-file-util.c`](https://github.com/git-for-windows/build-extra/blob/cad7897f73f1ac1810161772ce6c62986c9e1744/git-extra/Makefile#L7) and [installed like it](https://github.com/git-for-windows/build-extra/blob/cad7897f73f1ac1810161772ce6c62986c9e1744/git-extra/PKGBUILD#L117), too.

**Comment by sukiletxe** [comment]
@dscho I'm sorry, my C skills are non-existant right now.

**Comment by dscho** [comment]
@sukiletxe thank you for letting me know.

**Comment by dscho** [comment]
I opened https://github.com/git-for-windows/build-extra/pull/234 to address this, but first, @sukiletxe, I need confirmation that this `git-askyesno.exe` actually solves the problem.
